Temporal.Duration#toString should use BigInt#toString for values beyond MAX_SAFE_INTEGER

rkirsling opened the pull request:
#### 78b4ebbdfc351bea76965ea5562d5536d3783f17
<pre>
Temporal.Duration#toString should use BigInt#toString for values beyond MAX_SAFE_INTEGER
<a href="https://bugs.webkit.org/show_bug.cgi?id=243323">https://bugs.webkit.org/show_bug.cgi?id=243323</a>

Reviewed by Yusuke Suzuki.

Duration fields are currently specified as finite integral Numbers; unless a further bound is introduced in the future,
this means we&apos;ll need to use BigInt#toString for values beyond MAX_SAFE_INTEGER. (Until now we&apos;ve just being using 1e21
as a maximum, since this is the cutoff for Number#toFixed.) This is somewhat unfortunate since the use of BigInt incurs
a bunch of new exception logic.

Furthermore, TemporalDurationToString somehow expects us to display seconds as a decimal number in a case like
`Temporal.Duration.from({ microseconds: Number.MAX_VALUE, nanoseconds: Number.MAX_VALUE }).toString()`,
where calculating in the Number domain produces Infinity. I&apos;ve added this to my concerns in tc39/proposal-temporal#2195;
for now, I think it makes the most sense to implement this as a RangeError.

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::toString const):
(JSC::appendInteger):
(JSC::TemporalDuration::toString):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalDurationPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/252935@main">https://commits.webkit.org/252935@main</a>
</pre>

Constellation reviewed (approved):
r=me. And yeah, I think we should file a issue introducing a limit for this value. It does not make sense that we do not have a limit on duration's fields while actual time has limit.

rkirsling committed:
Temporal.Duration#toString should use BigInt#toString for values beyond MAX_SAFE_INTEGER
https://bugs.webkit.org/show_bug.cgi?id=243323

Reviewed by Yusuke Suzuki.

Duration fields are currently specified as finite integral Numbers; unless a further bound is introduced in the future,
this means we'll need to use BigInt#toString for values beyond MAX_SAFE_INTEGER. (Until now we've just being using 1e21
as a maximum, since this is the cutoff for Number#toFixed.) This is somewhat unfortunate since the use of BigInt incurs
a bunch of new exception logic.

Furthermore, TemporalDurationToString somehow expects us to display seconds as a decimal number in a case like
`Temporal.Duration.from({ microseconds: Number.MAX_VALUE, nanoseconds: Number.MAX_VALUE }).toString()`,
where calculating in the Number domain produces Infinity. I've added this to my concerns in tc39/proposal-temporal#2195;
for now, I think it makes the most sense to implement this as a RangeError.

* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/runtime/TemporalDuration.cpp:
(JSC::TemporalDuration::toString const):
(JSC::appendInteger):
(JSC::TemporalDuration::toString):
* Source/JavaScriptCore/runtime/TemporalDuration.h:
* Source/JavaScriptCore/runtime/TemporalDurationPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/252935@main
